### ISSUE TITLE
Extend stablebaseline 3 PPO implementation by with penalties

### ISSUE THREAD

**Issue by stheid** [stheid/constraint_water_treatment_gym#1]
In each policy update step, the penalty function is called with the policy and the current state. this results in a gradient.

Currently i have two ideas for such a penalty function, both need a (computationally) differentiable system approximation function f:
- calculate action a = policy(state), calculate succ-state = f(state,action), calculate penalty gradient on succ-state
- instead of taking the current state, the policy is evaluated on a predefined set of states. the rest simmilar

In order to use Pytorch autograd all functions need to be formulated in pytorch tensors, such that the policy can backpropagate from the loss

**Comment by stheid** [comment]
In my endever to reverse-engener the stable-baselines implementation i am currently stuck with the following things:
- what is the rollout-buffer?
- how do i get the current state (perhaps related to rollout-buffer)?

#### is PPO.train the function i want to overwrite?
Seems to be correct: https://stable-baselines3.readthedocs.io/en/master/modules/base.html#stable_baselines3.common.on_policy_algorithm.OnPolicyAlgorithm.train

**Comment by stheid** [comment]
So the rollout buffer contains a trajectory of states recorded with a "current policy". With this data we perform multiple updates until we collect new data with the next major policy.

This also answers the second question: the current state, needed for the policy update stems from the rollout buffer ans we can implement the train function as we like, using the buffered data multiple times (like done in their PPO implementation) or only use it once.
